lightning/src/ln/channelmanager.rs

```diff
@@ -18221,6 +18242,101 @@ where
 			}
 		}
 
+		// De-duplicate HTLCs that are present in both `failed_htlcs` and `decode_update_add_htlcs`.
+		// Omitting this de-duplication could lead to redundant HTLC processing and/or bugs.
+		for (src, _, _, _, _, _) in failed_htlcs.iter() {
+			if let HTLCSource::PreviousHopData(prev_hop_data) = src {
+				dedup_decode_update_add_htlcs(
+					&mut decode_update_add_htlcs,
+					prev_hop_data,
+					"HTLC was failed backwards during manager read",
+					&args.logger,
+				);
+			}
+		}
+
+		// See above comment on `failed_htlcs`.
+		for htlcs in claimable_payments.values().map(|pmt| &pmt.htlcs) {
+			for prev_hop_data in htlcs.iter().map(|h| &h.prev_hop) {
+				dedup_decode_update_add_htlcs(
+					&mut decode_update_add_htlcs,
+					prev_hop_data,
+					"HTLC was already decoded and marked as a claimable payment",
+					&args.logger,
+				);
+			}
+		}
+
+		// Remove HTLCs from `forward_htlcs` if they are also present in `decode_update_add_htlcs`.
```

Comment:
I appear to be missing where we remove HTLCs that were already forwarded? We remove items that were forwarded to closed channels, ones that were claimable/inbound to us, but I don't see where we remove elements that were forwarded to an open channel.